### PR TITLE
Pytorch visualizations

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR proposes to add visualizations of:

- intermediate activations
image

- intermediate layers
image

- model attention via [GradCam](https://arxiv.org/pdf/1610.02391.pdf)
image

### CODE REVIEW

**Comment by SergeyTsimfer** [comment]
A slight change to GradCam: now it allows us to pass any tensor to backpropagate from, for example, the true labels. This way we can get this image for seismic related tasks:
image

File: batchflow/models/torch/base.py [hunk 1]
```diff
@@ -1180,6 +1047,22 @@ def predict(self, *args, targets=None, feed_dict=None, train_mode='', fetches=No
         output = self._fill_output(fetches, output_container)
         return output
 
+    def _make_prediction_inputs(self, *args, targets=None, feed_dict=None, **kwargs):
+        """ Parse arguments to create valid inputs for the model. """
+        feed_dict = {**(feed_dict or {}), **kwargs}
+        if len(feed_dict) == 1:
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1053]: Add some comments on what's going on here

File: batchflow/models/torch/base.py [hunk 2]
```diff
@@ -1180,6 +1047,22 @@ def predict(self, *args, targets=None, feed_dict=None, train_mode='', fetches=No
         output = self._fill_output(fetches, output_container)
         return output
 
+    def _make_prediction_inputs(self, *args, targets=None, feed_dict=None, **kwargs):
+        """ Parse arguments to create valid inputs for the model. """
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1051]: Extend the docstring with examples of inputs and parsing results

File: batchflow/models/torch/base.py [hunk 3]
```diff
@@ -1343,3 +1227,349 @@ def load(self, path, *args, eval=False, **kwargs):
...
+        multiindex = pd.MultiIndex.from_tuples(indices, names=['iter', 'name'])
+
+        self.profile_info = pd.DataFrame(values, index=multiindex,
+                                         columns=['ncalls', 'CPU_tottime', 'CPU_cumtime', 'CUDA_cumtime'])
+        self.profile_info['CPU_tottime_avg'] = self.profile_info['CPU_tottime'] / self.profile_info['ncalls']
+        self.profile_info['CUDA_cumtime_avg'] = self.profile_info['CUDA_cumtime'] / self.profile_info['ncalls']
+
+
+    # Textual visualization of model
+    def information(self, config=True, devices=True, train_steps=True, model=False, misc=False):
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1288]: Make it as clean as possible. It should gather info into a readable structure and print only if needed (or no print at all)

File: batchflow/models/torch/base.py [hunk 4]
```diff
@@ -1343,3 +1227,349 @@ def load(self, path, *args, eval=False, **kwargs):
...
+            print(template.format('Last iteration params'))
+
+            iter_info = {**self.iter_info}
+            iter_info.pop('lr')
+            pprint(iter_info)
+
+    @property
+    def info(self):
+        """ Show information about model configuration, used devices, train steps and more. """
+        self.information()
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1336]: Property should not print!

File: batchflow/models/torch/base.py [hunk 5]
```diff
@@ -1343,3 +1227,349 @@ def load(self, path, *args, eval=False, **kwargs):
...
+
+
+    # Visualize intermediate layers: activations and features
+    def get_intermediate_activations(self, *args, layers=None, feed_dict=None, **kwargs):
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1340]: These are lo-o-o-ong methods and perhaps can be moved to a separate module

**Review by roman-kh — CHANGES_REQUESTED** [review]

File: batchflow/models/torch/visualization.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
+""" !!. """
```
Comment (roman-kh) [inline batchflow/models/torch/visualization.py:1]: ??
